python/tskit/trees.py

```diff
@@ -4901,10 +4883,16 @@ def alignments(
                 )
             ref_bytes = reference_sequence.encode("ascii")
             a[:] = np.frombuffer(ref_bytes, dtype=np.int8)
-        H = self._haplotypes_array(
-            isolated_as_missing=isolated_as_missing,
-            missing_data_character=missing_data_character,
-        )
+
+        # By rights we should detect this in the variants iterator in C,
+        # but in reality we'll not run this on any large tree sequences
+        # anyway, so lets not worry about the perf.
+        if any(tree.num_roots > 1 for tree in self.trees()):
```

Comment:
Is this the right way to test for missing data? This will fail e.g. on forward-simulated tree sequences, even though we might know the sequence. It will also fail if a sample is isolated but has a mutation above it: in this case the data is not usually thought of as missing.

Reply:
Yes, you're right, the forward simulation case would be incorrectly identified. I've pushed an update with as good a workaround as I can come up with, modulo actually sitting down and implementing the full thing in C.